src/main/java/net/earthcomputer/clientcommands/features/VillagerCracker.java

```diff
@@ -0,0 +1,469 @@
+            simulator.onBadRNG("potion");
+        } else if (!mc.player.getItemInHand(InteractionHand.MAIN_HAND).isEmpty()) {
+            simulator.onBadRNG("itemInMainHand");
+        } else {
+            Level level = targetVillager.level();
+            Vec3 pos = targetVillager.position();
+            BlockPos blockPos = targetVillager.blockPosition();
+
+            int villagersNearVillager = level.getEntities(EntityTypeTest.forExactClass(Villager.class), AABB.ofSize(pos, 10.0, 10.0, 10.0), entity -> entity.position().distanceToSqr(pos) <= 5.0 * 5.0).size();
+            if (villagersNearVillager > 1) {
+                simulator.onBadRNG("gossip");
+                return;
+            }
+
+            List<BlockPos> validBedHeadPositions = List.of(blockPos.north(3), blockPos.east(3), blockPos.south(3), blockPos.west(3));
+            List<BlockPos> bedHeadPositions = BlockPos.withinManhattanStream(blockPos, 15, 7, 15).map(BlockPos::new).filter(p -> level.getBlockState(p).is(BlockTags.BEDS) && level.getBlockState(p).getValue(BedBlock.OCCUPIED) == Boolean.FALSE && level.getBlockState(p).getValue(BedBlock.PART) == BedPart.HEAD).toList();
+            Direction bedDirection;
+            if (bedHeadPositions.size() == 1 && validBedHeadPositions.contains(bedHeadPositions.getFirst())) {
+                bedDirection = Direction.Plane.HORIZONTAL.stream().skip(validBedHeadPositions.indexOf(bedHeadPositions.getFirst())).findAny().orElse(null);
+            } else {
+                simulator.onBadRNG("invalidBedPosition");
+                sendInvalidSetupHelp();
+                return;
+            }
+
+            for (Direction direction : Direction.Plane.HORIZONTAL) {
+                BlockPos airPos = blockPos.relative(direction);
+                BlockPos trapdoorPos = airPos.above();
+                BlockState airPosState = level.getBlockState(airPos);
+                BlockState trapdoorPosState = level.getBlockState(trapdoorPos);
+                if (!((airPosState.isAir() || direction != bedDirection) && trapdoorPosState.is(BlockTags.TRAPDOORS) && trapdoorPosState.getValue(TrapDoorBlock.HALF) == Half.TOP && trapdoorPosState.getValue(TrapDoorBlock.HALF) == Half.TOP && trapdoorPosState.getValue(TrapDoorBlock.OPEN) == Boolean.FALSE)) {
+                    simulator.onBadRNG("invalidCage");
+                    sendInvalidSetupHelp();
+                    return;
+                }
+            }
+
+        }
+    }
+
+    private static void sendInvalidSetupHelp() {
+        ClientCommandHelper.sendHelp(Component.translatable("villagerManip.help.setup.prefix").append(Component.translatable("villagerManip.help.setup.here").withStyle(ChatFormatting.UNDERLINE).withStyle(style -> style.withClickEvent(new ClickEvent(ClickEvent.Action.OPEN_URL, "https://raw.githubusercontent.com/Earthcomputer/clientcommands/refs/heads/fabric/villager_rng_setup.png")))).append(Component.translatable("villagerManip.help.setup.suffix")));
```

Comment:
I think this should be a permalink, otherwise we're committing to having this file in the repository forever. The problem is, we don't have a permalink pointing to my repo yet. I'll add something to the clientcommands buildscript that allows you to get the commit hash that clientcommands was built with

Reply:
ok